### PR TITLE
Configuring App icons via app.json

### PR DESCRIPTION
**Pull request by ndejesus1227**
The problem: Setting up app icons on both iOS and Android is very tedious and is quite a time sink. The solution is to drop your app icon image at root level and use app.json property "icon": "yourAppIcon.png" to automatically handle everything for you by creating different sized images and placing them in the appropriate folders and configuring contents.json for iOS.

Test plan is to do as I instructed above and ensure that the icon will be displayed as needed.

Video demonstration: https://www.youtube.com/watch?v=5DAgi2TYAJ0

### CODE REVIEW

File: local-cli/eject/eject.js [hunk 1]
```diff
@@ -64,6 +66,15 @@ function eject() {
     process.exit(1);
   }
 
+  const appIcon = appConfig.icon;
+  if(!appIcon) {
```
Comment (ericvicenti) [inline local-cli/eject/eject.js]: The formatting is inconsistent, should be `if (`

The app icon should not be mandatory, we should have a fallback icon

File: local-cli/eject/eject.js [hunk 2]
```diff
@@ -84,6 +129,29 @@ function eject() {
       appName,
       templateOptions
     );
+
+    console.log('Setting up App Icons');
+    const hdpiPath = 'android/app/src/main/res/mipmap-hdpi/ic_launcher.png';
```
Comment (ericvicenti) [inline local-cli/eject/eject.js]: Does this work on windows? I think this path separator may not work there.

Its not great that the path is duplicated several times here

File: local-cli/eject/eject.js [hunk 3]
```diff
@@ -84,6 +129,29 @@ function eject() {
...
+    const mdpiPath = 'android/app/src/main/res/mipmap-mdpi/ic_launcher.png';
+    const xhdpiPath = 'android/app/src/main/res/mipmap-xhdpi/ic_launcher.png';
+    const xxhdpiPath = 'android/app/src/main/res/mipmap-xxhdpi/ic_launcher.png';
+
+    const pictureSizes = [48, 72, 96, 144];
```
Comment (ericvicenti) [inline local-cli/eject/eject.js]: Maybe we could express this in a clearer data structure.

```
const ICON_SIZES = {
  android_hdpi: {
    path: 'android/app/src/main/res/mipmap-hdpi/ic_launcher.png'
    size: 48,
  },
  ...
];
```

File: local-cli/eject/eject.js [hunk 4]
```diff
@@ -72,8 +83,42 @@ function eject() {
...
+      appIcon,
       templateOptions
     );
+
+    console.log('Setting Up App Icons.')
+    const pictureSizes = [40, 60, 58, 87, 80, 120, 180];
+    let contentPath = 'ios/SomeTest/Images.xcassets/AppIcon.appiconset/Contents.json';
+
+    let fileNames = pictureSizes.reduce( (acc, size) => {
+      let filePath = `ios/SomeTest/Images.xcassets/AppIcon.appiconset/${size}pt-${appIcon}`;
```
Comment (ericvicenti) [inline local-cli/eject/eject.js]: Always use `const` when a variable doesn't get re-assigned.

File: local-cli/eject/eject.js (hunk 1 above)
Comment (ndejesus1227) [inline local-cli/eject/eject.js]: Noted, will change tomorrow

File: local-cli/eject/eject.js (hunk 2 above)
Comment (ndejesus1227) [inline local-cli/eject/eject.js]: I haven't tried this on Windows... I can investigate

File: local-cli/eject/eject.js (hunk 3 above)
Comment (ndejesus1227) [inline local-cli/eject/eject.js]: Sounds good to me! Will work on this during the weekend

**Comment by chirag04** [comment]
Might be worth making it a cli command. react-native generate-icons path-to-icon

File: local-cli/eject/eject.js (hunk 1 above)
Comment (ndejesus1227) [inline local-cli/eject/eject.js]: I decided to get rid of this whole if statement that makes the icon mandatory. If there is no icon it'll fall back to what it does already without this configuration

**Comment by ericvicenti** [comment]
I think it should not be an independent command from eject. If the native code is already pre-generated, how does the generator know how to install the icon correctly?

**Comment by chirag04** [comment]
generating app icons can happen at any point and not just when ejecting. so i thought it would make sense to have it as an independent command.

> If the native code is already pre-generated, how does the generator know how to install the icon correctly

we can take the path to `AppIcon.appiconset` as an argument. likewise on android.

my motivation here is that having it as an independent command will serve a boarder audience.

File: local-cli/eject/eject.js (hunk 2 above)
Comment (janicduplessis) [inline local-cli/eject/eject.js]: I think it should work, we're just passing these paths to the sharp library and it should resolve it properly, forward slashes are valid for paths on windows.

File: local-cli/eject/eject.js (hunk 3 above)
Comment (ndejesus1227) [inline local-cli/eject/eject.js]: @ericvicenti I just updated it with your data structure proposal. Lemme know if this works for you!

**Comment by ndejesus1227** [comment]
I think @chirag04 has a great point. I feel like it would make sense if we can utilize app.json without the eject command. Not to say that we take it out of eject, but provide another way to do it. It was pretty tedious to have to delete android and iOS folder just to see this work. 

What do you think @ericvicenti ?

**Comment by ericvicenti** [comment]
I think it would be nice if icons could be overridden. In `app.json`:

```
{
  "icon": {
    "android": "./myAndroidIcon.png",
    "android_hidpi": "./myAndroidOverrideIcon.png",
    "ios_120": "./mySpecialIOSIcon.png",
    "default": "./myIcon.png",
}
```

The conventions could change for this, but I expect it will be a common requirement for real-world apps to specify a different icon for iOS as Android.

If we had a command line, how would we support an advanced configuration like this?

I'm also worried about having too many commands to maintain. There are going to be lots of other things in the `app.json` file that is used during the eject step:

- launch screen
- permissions config and strings
- keyboard config (android)
- linking setup
- bundle identifier

Are we going to have a `react-native` command for all of these? That feels like a very high maintenance burden.

**Comment by ericvicenti** [comment]
In other words, I would much rather see you move on to other features like splash image overriding, rather than spending time on extra CLI commands

**Comment by ndejesus1227** [comment]
@ericvicenti makes total sense! So how does this sound:

I add changes so that you can specify between iOS and Android icons only for now and I move on to figuring out the splash page through app.json? 

Also, I was wondering about documentation for react-native eject. Would it be too early to start fleshing that out?

**Comment by ndejesus1227** [comment]
@ericvicenti so because you said that those naming conventions may change for overriding app icons, I did the bare minimum of at least making it possible to set different icons for iOS and Android. Let me know what you think!

I want to work on the splash page, should that be a separate pull request?

File: local-cli/eject/eject.js [hunk 5]
```diff
@@ -74,6 +80,42 @@ function eject() {
       appName,
       templateOptions
     );
+
+    console.log('Setting Up App Icons.')
+    if(!appIconiOS) {
+      appIconiOS = appIcon;
+    }
+    const pictureSizes = [40, 60, 58, 87, 80, 120, 180];
+    const contentPath = 'ios/SomeTest/Images.xcassets/AppIcon.appiconset/Contents.json';
```
Comment (ericvicenti) [inline local-cli/eject/eject.js]: What is `SomeTest`, and does it need to be hardcoded in this path?

File: local-cli/eject/eject.js [hunk 6]
```diff
@@ -84,6 +126,54 @@ function eject() {
       appName,
       templateOptions
     );
+
+    console.log('Setting up App Icons');
```
Comment (ericvicenti) [inline local-cli/eject/eject.js]: can the log mention the platform of the icons being set up? Same with the iOS log

File: local-cli/eject/eject.js [hunk 7]
```diff
@@ -84,6 +126,54 @@ function eject() {
...
+      android_xxhdpi: {
+        path: 'android/app/src/main/res/mipmap-xxhdpi/ic_launcher.png',
+        size: 144
+      }
+    }];
+
+    Object.keys(ICON_SIZES[0]).forEach((prop) => {
+      const path = ICON_SIZES[0][prop].path;
+      const size = ICON_SIZES[0][prop].size;
+      console.log(path, size);
```
Comment (ericvicenti) [inline local-cli/eject/eject.js]: can you remove it or add text around it to be more readable?

File: local-cli/eject/eject.js [hunk 8]
```diff
@@ -64,6 +66,10 @@ function eject() {
     process.exit(1);
   }
 
+  let appIconiOS = appConfig.icon.ios;
+  let appIconAndroid = appConfig.icon.android;
+  const appIcon = appConfig.icon.default;
```
Comment (ericvicenti) [inline local-cli/eject/eject.js]: Can we handle the default behavior up here?

```
const iconConfig = (typeof appConfig.icon === 'string' ? {default: appConfig.icon} : appConfig.icon  ) || {};
const appIconIOS = iconConfig.ios || iconConfig.defaultIcon;
```

File: local-cli/eject/eject.js (hunk 5 above)
Comment (ndejesus1227) [inline local-cli/eject/eject.js]: Oops! that's what I named the project with app.json, I should make that dynamic. Thank you

File: local-cli/eject/eject.js (hunk 7 above)
Comment (ndejesus1227) [inline local-cli/eject/eject.js]: I'm sorry, remove what exactly? That console.log?

**Comment by ndejesus1227** [comment]
@ericvicenti I may need to put some more thought on where the eject document goes, but I have an idea:

Under "Reference" from the main page of the React docs, perhaps there can be a document for "React Config" where it would detail the purpose of `react-native eject` and how to utilize app.json for setting up your app icons, splash page and other things that you've expressed you had plans for. Figured it should go there since using eject isn't necessary but can be very useful.

I personally haven't seen any bad situations with my experience using eject, I can't think of any edge cases or warnings for anything that can go wrong for now. 

What do you think? Gimme the word and I'll figure out how to start the React Config docs under "Reference" and talk about eject/app.json :)

**Comment by ncuillery** [comment]
The `eject` command is quite similar as _running `react-native upgrade` and answering Yes to each file changed_. `eject` improves this process, thanks to the `app.json` file.

Both `eject` and `upgrade` commands are relative to "bringing changes in the native apps". How about putting the doc into the current [Upgrading guide](https://facebook.github.io/react-native/docs/upgrading.html)?

@mkonicek and I aren't sure about replacing the current `upgrade` command with the `react-native-git-upgrade` implementation. Maybe we could considered that `eject` is a better replacement for the `upgrade` command. That's what people are looking for (see #12261).

Replacing the [2nd part of the Upgrading guide](https://facebook.github.io/react-native/docs/upgrading.html#alternative) (the "upgrade" command section) with the `eject` command documentation makes sense to me. We could also update the deprecation [warning](https://github.com/facebook/react-native/blob/master/local-cli/upgrade/upgrade.js#L35-L43) of the `upgrade` command as something like "You can use either `react-native eject` to upgrade the template and relink your librairies afterward or `react-native-git-upgrade` to merge the new template automatically".

WYT ?

**Comment by ndejesus1227** [comment]
@ericvicenti Hey, I added some changes and tested cases with icon images and without and everything works as expected. No icons default to to what RN does without the eject command and adding icons over writes that.

Let me know what you think of my updates and I'll incorporate your feedback and get started on documentation with your blessing.

**Comment by ndejesus1227** [comment]
@ncuillery I just looked over the upgrade docs. Correct me if I'm wrong, but I think upgrade does a bit more than eject. In the docs it says:

"Upgrading to new versions of React Native will give you access to more APIs, views, developer tools and other goodies."

I think the goal with eject is to rebuild your projects if things have either gone wrong on the native files and also serve as a way of reconfiguring certain parts of your project (so far all it does is allow you to change the name of the project and your app icons).

With that being said, it makes sense to me to put eject documentation on the upgrade docs since they both, as you said, "brings changes to the native apps".

I'm going to get started on the eject docs. What do you think?

**Comment by ncuillery** [comment]
@ndejesus1227 the `upgrade` command is basically a call to `copyProjectTemplateAndReplace` (see https://github.com/facebook/react-native/blob/master/local-cli/upgrade/upgrade.js#L141), it "upgrades" React Native because the user is supposed to run `npm i react-native@latest` before 😄

**Comment by ndejesus1227** [comment]
@ncuillery oh!!! that makes sense, thanks for clearing that up. Still waiting on word from eric, but I started the docs for eject already. Can I get your feedback on it?

https://gist.github.com/ndejesus1227/0eb673452c34e085b3f04fd5c1cbcc9f

**Comment by ericvicenti** [comment]
For now, lets just add the icon docs to the top of the eject command file. See on line 25 where you can add the explanation of the icon field. Can you update it, then we can land this PR?

We can create improved upgrade docs as a followup PR

**Comment by ndejesus1227** [comment]
@ericvicenti I've added comments to the eject file, is that good enough? I'm now trying to figure out how to pass the travis and circle ci tests

File: local-cli/eject/eject.js [hunk 9]
```diff
@@ -21,6 +23,10 @@ const fs = require('fs');
  *
  * - `name` - The short name used for the project, should be TitleCase
  * - `displayName` - The app's name on the home screen
+ * - `icon` - with three other properties, `ios`, `android`, `default`
```
Comment (ericvicenti) [inline local-cli/eject/eject.js]: I think it would be good to start by explaining that icon can be set to a path string, and that will be used as the icon on every platform

For the deep configuration, please create an indented sub-list of options.

File: local-cli/eject/eject.js [hunk 10]
```diff
@@ -64,6 +70,10 @@ function eject() {
     process.exit(1);
   }
 
+  const iconConfig = (typeof appConfig.icon === 'string' ? {default: appConfig.icon.default} : appConfig.icon  ) || {};
```
Comment (ericvicenti) [inline local-cli/eject/eject.js]: I believe this should be changed to:

```
const iconConfig = (typeof appConfig.icon === 'string' ? {default: appConfig.icon} : appConfig.icon  ) || {};
```

For each PR to react native, we also expect to see a test plan, that would cover the edge cases of the changes you're making. This is one edge case that I would expect to be in the test plan, and hopefully it would help surface the issue sooner

**Comment by ndejesus1227** [comment]
@ericvicenti are the changes to the documenting part good enough for you? I'm not exactly sure how to document edge cases. Should it be a video or screen shots or something?

**Comment by ericvicenti** [comment]
Did you test the `eject` step with a missing icon? What about an icon set with a bare string? what about icon with default and nothing else? What about an icon with a default and one platform set? Does everything work well when there is no default, and you only need an icon for one platform?

These are the edge cases I'd expect you to test. If everything is working then I'd say this change is ready to land!

**Comment by ndejesus1227** [comment]
@ericvicenti oh yes! I actually tested all of those cases two weeks ago and everything seemed smooth. Let me do it all again to double check, I'll document these results. Also, is it excessive that I tag you with each time I comment? Do you get the notifications regardless? lol

**Comment by ericvicenti** [comment]
Its not excessive because I get a ton of GH notifs and sometimes they fall through the cracks. Tagging me does help

**Comment by ndejesus1227** [comment]
@ericvicenti Hey Eric! I tested my code and I actually came across an issue! I made a fix for it in my recent commit if you don't mind checking it out.

I documented my findings here: https://gist.github.com/ndejesus1227/d5a80106cb8970d17749b2024343866e

Just some info on the issue I had. I refined the Android part of the app icon set up to check if there is a string that comes from the `appIconAndroid` constant AND make sure that the file that's being referred to exists. It only checks for images at root level, though. I think that's fine personally because you can drop those images at root level, run `eject` to set up the app icons and remove them after.

Other than that, all edge cases I can think of passed, if you think of any more I'd be happy to test.

File: local-cli/eject/eject.js [hunk 11]
```diff
@@ -84,6 +129,54 @@ function eject() {
       appName,
       templateOptions
     );
+
+    console.log('Setting up App Icons for Android');
+    const hdpiPath = 'android/app/src/main/res/mipmap-hdpi/ic_launcher.png';
```
Comment (ericvicenti) [inline local-cli/eject/eject.js]: It looks like this is duplicated with ICON_SIZES below. Can you refactor it so that these strings are not duplicated?

**Comment by ndejesus1227** [comment]
@ericvicenti I refactored. Let me know if that's what you had in mind.

**Comment by itinance** [comment]
Any news on this? Wanna see this in RN 0.44  :)

**Comment by ndejesus1227** [comment]
@ericvicenti hey man, any update? Is there anything I can do to get my tests to pass at least? I also started my next pull request which will do the same for the splash page

@itinance  thanks for your support :D

**Comment by ericvicenti** [comment]
Ah, we ran into a snag with adding the `sharp` dependency. I'll try to get this landed ASAP

Thanks for the contribution, @ndejesus1227!

File: local-cli/eject/eject.js [hunk 12]
```diff
@@ -84,6 +129,53 @@ function eject() {
...
+        })
+      })
+    };
+
+    Object.keys(ICON_SIZES[0]).forEach((prop) => {
+      const path = ICON_SIZES[0][prop].path;
+      const size = ICON_SIZES[0][prop].size;
+      sharp(appIconAndroid)
+      .resize(size, size)
+      .toFile(path), function(err) {
```
Comment (ericvicenti) [inline local-cli/eject/eject.js]: wait, this looks like a typo. What is this function doing here? What happens when you run this?

Comment (ndejesus1227) [inline local-cli/eject/eject.js]: I don't see a typo? This uses sharp to spit out an image an image of the same width and height to whatever path it is on in the ICON_SIZES object. It works as expected

Comment (jamesisaac) [inline local-cli/eject/eject.js]: @ndejesus1227 The closing parenthesis from `toFile(path)` is in the wrong place (I'm guessing you meant to pass in the error handler as 2nd argument).  Right now that comma is in the middle of nowhere and invalid JS syntax.

Comment (ndejesus1227) [inline local-cli/eject/eject.js]: @jamesisaac doh! Thank you. I just fixed it!

@ericvicenti is there anything I can do to make sure it passes the tests or is that all internal? The sharp dependency seems beyond my reach

**Comment by ndejesus1227** [comment]
@ericvicenti any update?

**Comment by itinance** [comment]
*sigh*

**Comment by ndejesus1227** [comment]
@ericvicenti  perhaps there is someone else that can investigate the sharp dependency issue?

**Comment by ndejesus1227** [comment]
@ericvicenti A friendly nudge

**Comment by ndejesus1227** [comment]
@ericvicenti hoping for an update

**Comment by ericvicenti** [comment]
Hey @ndejesus1227, sorry for the long wait! I've talked about it with some other folks at FB and it seems that the team has been moving RN away from native dependencies, and would not want to add a new one now.

So unfortunately it looks like we need to do this another way. Perhaps we can make a separate library that "ejects" and handles the icon? If we do it outside this repo we can move faster because this code is synced to FB, who does not need the eject functionality.

Perhaps it could work like this:

```
npm install --save-dev react-native-eject
# edit package.json, add "react-native-eject" to `scripts.eject`
# edit app.json to configure icon and native code
rm -rf ios android
npm run eject
# the module will generate the native code with `react-native eject`, then set up the icon
```

Want to take a stab at implementing this module? Otherwise I can prototype it at some point in the next few weeks.

**Comment by ndejesus1227** [comment]
@ericvicenti Thank you eric! No issues, I know you have a lot on your plate :)

I would LOVE an opportunity to kick this off. I'm going to need some time to take a stab at it. May I nudge you for guidance/feedback every now and then? What would be the best way for me to contact you if so?

Thank you!!!!

**Comment by ericvicenti** [comment]
I'm glad you're excited! Github issues are probably the best place to discuss. If you tag me I'll be more likely to see it and respond relatively quickly :-D

**Comment by hramos** [comment]
It seems like this PR is not likely to be merged, so I'll go ahead and close it.

**Comment by itinance** [comment]
Sadly... it would be a nice feature

**Comment by ndejesus1227** [comment]
@hramos That's fine, I am still planning on making a separate module for this, just have my hands tied with another project.


@itinance  I'm not giving up! Don't you worry :p
